packages/flutter/lib/src/widgets/selection_gestures.dart

```diff
@@ -0,0 +1,1320 @@
+  // or the recognizer is disposed of then this value is reset.
+  Set<LogicalKeyboardKey> get keysPressedOnDown => _keysPressedOnDown ?? <LogicalKeyboardKey>{};
+  // Whether the tap drifted past the tolerance defined by [kDoubleTapTouchSlop] in any subsequent
+  // tracked [PointerMoveEvent]s.
+  //
+  // This value default to false.
+  //
+  // If the tap does drift past the tolerance then we reset all of the tracked state except
+  // the [currentDown], [currentUp], [consecutiveTapCount], and [keysPressedOnDown]. This is because
+  // the [OneSequenceGestureRecognizer] may be handling a gesture that does accept a tap drift past
+  // the tolerance defined by [kDoubleTapTouchSlop], such as a drag, so it may still want access
+  // to the tracked tap state.
+  bool get pastTapTolerance => _pastTapTolerance;
+  // The upper limit for the [consecutiveTapCount]. When this limit is reached
+  // all tap related state is reset and a new tap series is tracked.
+  //
+  // If this value is null, [consecutiveTapCount] can grow infinitely large.
+  int? get upperLimit;
+
+  // Private tap state tracked.
+  PointerDownEvent? _down;
+  PointerUpEvent? _up;
+  int _consecutiveTapCount = 0;
+  Set<LogicalKeyboardKey>? _keysPressedOnDown;
+  bool _pastTapTolerance = false;
+
+  bool _wonArena = false;
+  OffsetPair? _originPosition;
+  int? _previousButtons;
+
+  // For timing taps.
+  Timer? _consecutiveTapTimer;
+  Offset? _lastTapOffset;
+
+  // When we start to track a tap, we can choose to increment the [consecutiveTapCount]
+  // if the given tap falls under the tolerance specifications or we can reset the count to 1.
+  @override
+  void addAllowedPointer(PointerDownEvent event) {
+    super.addAllowedPointer(event);
+    if (upperLimit != null && upperLimit == _consecutiveTapCount) {
+      _tapTrackerReset();
+    }
+    _up = null;
+    _pastTapTolerance = false;
+    _wonArena = false;
+    if (_down != null && !_representsSameSeries(event)) {
+      // The given tap does not match the specifications of the series of taps being tracked,
+      // reset the tap count and related state.
+      _consecutiveTapCount = 1;
+    } else {
+      _consecutiveTapCount += 1;
+    }
+    _consecutiveTapTimerStop();
+    // `_down` must be assigned in this method instead of [handleEvent],
+    // because [acceptGesture] might be called before [handleEvent],
+    // which may rely on `_down` to initiate a callback.
+    _trackTrap(event);
+  }
+
+  @override
+  void acceptGesture(int pointer) {
+    _wonArena = true;
+    if (_up != null && _down != null) {
+      _consecutiveTapTimerStop();
+      _consecutiveTapTimerStart();
+      _wonArena = false;
+    }
+  }
+
+  double _getGlobalDistance(PointerEvent event) {
+    assert(_originPosition != null);
+    final Offset offset = event.position - _originPosition!.global;
+    return offset.distance;
+  }
+
+  @override
+  void handleEvent(PointerEvent event) {
+    if (event is PointerMoveEvent) {
```

Comment:
Why does this need to check for move? we should only need to worry about the slop distance when the pointer is up?

Reply:
There can be a `PointerMoveEvent` that occurs between a `PointerDownEvent` and `PointerUpEvent`. See https://github.com/flutter/flutter/blob/52b3dc25f6471c27b2144594abb11c741cb88f57/packages/flutter/lib/src/gestures/recognizer.dart#L611